src/util/rocksdb/properties.rs

```diff
@@ -466,6 +467,188 @@ impl DecodeProperties for UserCollectedProperties {
+    }
+
+    pub fn get_approximate_keys_in_range(&self, start: &[u8], end: &[u8]) -> u64 {
+        self.get_approximate_distance_in_range(OffsetKindInRange::Keys, start, end)
+    }
+
+    fn get_approximate_distance_in_range(
+        &self,
+        kind: OffsetKindInRange,
+        start: &[u8],
+        end: &[u8],
+    ) -> u64 {
+        assert!(start <= end);
+        if start == end {
+            return 0;
+        }
+
+        let range = self.offsets.range::<[u8], _>((Unbounded, Included(start)));
+        let start_offset = match range.last() {
+            Some((_, v)) => v.get(kind),
+            None => 0,
+        };
+        let mut range = self.offsets.range::<[u8], _>((Included(end), Unbounded));
+        let end_offset = match range.next() {
+            Some((_, v)) => v.get(kind),
+            None => self.offsets.iter().last().map_or(0, |(_, v)| v.get(kind)),
+        };
+        if end_offset < start_offset {
+            panic!(
+                "start {:?} end {:?} start_offset {} end_offset {}",
+                start, end, start_offset, end_offset
+            );
+        }
+        end_offset - start_offset
+    }
+
+    pub fn smallest_key(&self) -> Option<Vec<u8>> {
+        self.offsets.iter().next().map(|(key, _)| key.clone())
+    }
+
+    pub fn largest_key(&self) -> Option<Vec<u8>> {
+        self.offsets.iter().last().map(|(key, _)| key.clone())
+    }
+}
+
+impl From<SizeProperties> for RangeProperties {
+    fn from(p: SizeProperties) -> RangeProperties {
+        let mut res = RangeProperties::default();
+        for (key, size_handle) in p.index_handles.0 {
+            let mut range = OffsetsInRange::default();
+            range.size = size_handle.offset;
+            res.offsets.insert(key, range);
+        }
+        res
+    }
+}
+
+#[derive(Default)]
+pub struct RangePropertiesCollector {
+    props: RangeProperties,
+    last_offsets: OffsetsInRange,
+    last_key: Vec<u8>,
+    cur_offsets: OffsetsInRange,
+}
+
+impl RangePropertiesCollector {
+    fn size_in_last_range(&self) -> u64 {
+        self.cur_offsets.size - self.last_offsets.size
+    }
+
+    fn keys_in_last_range(&self) -> u64 {
+        self.cur_offsets.keys - self.last_offsets.keys
+    }
+
+    fn insert_new_point(&mut self, key: Vec<u8>) {
+        self.last_offsets = self.cur_offsets.clone();
+        self.props.offsets.insert(key, self.cur_offsets.clone());
+    }
+}
+
+impl TablePropertiesCollector for RangePropertiesCollector {
+    fn add(&mut self, key: &[u8], value: &[u8], entry_type: DBEntryType, _: u64, _: u64) {
+        if entry_type != DBEntryType::Put {
+            return;
+        }
+
+        // keys
+        self.cur_offsets.keys += 1;
+        // size
+        let size = key.len() + value.len();
+        self.cur_offsets.size += size as u64;
+        // Add the start key for convenience.
+        if self.last_key.is_empty() || self.size_in_last_range() >= PROP_SIZE_INDEX_DISTANCE {
```

Comment:
We need to add a `PROP_KEYS_INDEX_DISTANCE` and record a point whenever `size >= PROP_SIZE_INDEX_DISTANCE || keys >= PROP_KEYS_INDEX_DISTANCE`. Otherwise, for small entries, a range will contain too many keys.